crates/matrix-sdk/src/widget_api/driver/mod.rs

```diff
@@ -0,0 +1,231 @@
+use self::widget::Widget;
+use super::{
+    handler::{self, Capabilities, OpenIDState},
+    messages::{
+        capabilities::{Filter, Options, EventFilter},
+        from_widget::{ReadEventRequest, ReadEventResponse, SendEventRequest, SendEventResponse},
+        {openid, MatrixEvent},
+    },
+    {Error::WidgetError, Result},
+};
+use crate::room::MessagesOptions;
+use crate::{event_handler::EventHandlerHandle, room::Joined};
+use async_trait::async_trait;
+use ruma::{api::client::filter::RoomEventFilter, events::AnySyncTimelineEvent, serde::Raw};
+use tokio::sync::mpsc;
+
+pub mod widget;
+
+#[derive(Debug)]
+pub struct Driver<W: Widget> {
+    pub matrix_room: Joined,
+    pub widget: W,
+    add_event_handler_handle: Option<EventHandlerHandle>,
+}
+
+#[async_trait(?Send)]
+impl<W: Widget> handler::Driver for Driver<W> {
+    async fn send(&self, _message: handler::Outgoing) -> Result<()> {
+        // let message_str = serde_json::to_string(&message)?;
+        let _ = self.widget.send_widget_message("TODO get message string from outgoing");
+        Result::Ok(())
+    }
```

Comment:
This is superseeded by your send function now:
```rust
[async_trait]
impl<T: WidgetClient> HandlerDriver for Driver<T> {
    async fn send<M: OutgoingMessage>(&self, msg: M) -> Result<M::Response> {
        let id = Uuid::new_v4();
        let request_id = id.to_string();
        let raw = msg.into_message(Header { request_id, widget_id: self.id.clone() });
        self.sink.send(raw).map_err(|_| Error::WidgetDied)?;

        let (tx, rx) = oneshot::channel();
        self.pending.lock().expect("Pending mutex poisoned").insert(id.to_string(), tx);
        Ok(M::extract_response(rx.await.map_err(|_| Error::WidgetDied)?)?)
    }
}
```
It was very much a placeholder for this.

Reply:
I keep it in there so i dont have to update handler/mod.rs and introduce additional merge conflicts. This will be marked with an unused warning as soon as we merge your pr (I think). I also added unimpleented so that you it is clear that we can remove it once we have your alternative implementation